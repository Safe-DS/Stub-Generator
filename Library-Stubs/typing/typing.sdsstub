package typing

class Mapping

class Sequence
