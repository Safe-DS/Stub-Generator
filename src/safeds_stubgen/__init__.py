"""Safe-DS stubs generator."""
from __future__ import annotations
