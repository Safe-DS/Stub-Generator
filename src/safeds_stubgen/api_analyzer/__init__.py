"""API-Analyzer for the Safe-DS stubs generator."""
from __future__ import annotations

from ._api import API, Attribute, Class, Function, Parameter, ParameterAssignment
from ._get_api import get_api
from ._mypy_helpers import get_classdef_definitions, get_funcdef_definitions, get_mypyfile_definitions
from ._package_metadata import distribution, distribution_version, package_root
from ._types import (
    AbstractType,
    BoundaryType,
    DictType,
    EnumType,
    FinalType,
    ListType,
    LiteralType,
    NamedType,
    OptionalType,
    SetType,
    TupleType,
    UnionType,
)

__all__ = [
    "AbstractType",
    "API",
    "Attribute",
    "BoundaryType",
    "Class",
    "DictType",
    "distribution",
    "distribution_version",
    "EnumType",
    "FinalType",
    "Function",
    "get_api",
    "get_classdef_definitions",
    "get_funcdef_definitions",
    "get_mypyfile_definitions",
    "ListType",
    "LiteralType",
    "NamedType",
    "OptionalType",
    "package_root",
    "Parameter",
    "ParameterAssignment",
    "SetType",
    "TupleType",
    "UnionType",
]
