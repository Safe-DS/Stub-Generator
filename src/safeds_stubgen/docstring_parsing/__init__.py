"""Parsing docstrings into a common format."""

from ._abstract_docstring_parser import AbstractDocstringParser
from ._create_docstring_parser import create_docstring_parser
from ._docstring import AttributeDocstring, ClassDocstring, FunctionDocstring, ParameterDocstring, ResultDocstring
from ._docstring_style import DocstringStyle
from ._epydoc_parser import EpydocParser
from ._googledoc_parser import GoogleDocParser
from ._numpydoc_parser import NumpyDocParser
from ._plaintext_docstring_parser import PlaintextDocstringParser
from ._restdoc_parser import RestDocParser

__all__ = [
    "AbstractDocstringParser",
    "AttributeDocstring",
    "ClassDocstring",
    "create_docstring_parser",
    "DocstringStyle",
    "EpydocParser",
    "FunctionDocstring",
    "GoogleDocParser",
    "NumpyDocParser",
    "ParameterDocstring",
    "PlaintextDocstringParser",
    "RestDocParser",
    "ResultDocstring",
]
