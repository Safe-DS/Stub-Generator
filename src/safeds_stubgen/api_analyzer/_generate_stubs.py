from __future__ import annotations

from pathlib import Path
from typing import Any, TextIO


# Todo: Wir ignorieren self parameter im Init, also sollten wir für __init__ Funktionen is_self für parameter
#  überprüfen und diesen als Namen dann immer "self" geben, damit diese bei der Classen Erstellung ignoriert werden
#  können
# Todo: In der api.json output Datei alle Listen zu dictionaries, da wir sonst immer durch alle Einträge iterieren
#  müssen
class StubsGenerator:
    api_data: dict[str, Any]
    out_path: Path

    def __init__(self, api_data: dict[str, Any], out_path: Path) -> None:
        self.api_data = api_data
        self.out_path = out_path

    def generate_stubs(self) -> None:
        create_directory(Path(self.out_path / self.api_data["package"]))
        self._create_module_files()

    # Todo Handle __init__ files
    #  Dont create modules that start with "_" unless they are reexported with an alias without "_"
    #  or contain a reexported or public class (?)
    def _create_module_files(self) -> None:
        modules = self.api_data["modules"]

        for module in modules:
            if module["name"] == "__init__":
                continue

            module_dir = Path(self.out_path / module["id"])
            create_directory(module_dir)

            file_path = Path(self.out_path / module["id"] / f"{module["name"]}.sdsstub")
            create_file(file_path)

            with file_path.open("w") as f:
                # Create package info
                package_info = module["id"].replace("/", ".")
                f.write(f"package {package_info}\n\n")

                # Create imports
                self._write_qualified_imports(f, module["qualified_imports"])
                self._write_wildcard_imports(f, module["wildcard_imports"])
                f.write("\n\n")

                # Create global functions
                for function in self.api_data["functions"]:
                    if function["is_public"] and function["id"] in module["functions"]:
                        function_string = self._create_function_string(function)
                        f.write(f"{function_string}\n\n")

                # Create classes, class attr. & class methods
                for class_ in self.api_data["classes"]:
                    if class_["is_public"] and class_["id"] in module["classes"]:
                        self._write_class(f, class_)

                # Create enums & enum instances
                for enum in self.api_data["enums"]:
                    if enum["id"] in module["enums"]:
                        self._write_enum(f, enum)

    # Todo assigned_by, constructors, subclassing
    def _write_class(self, f: TextIO, class_data: dict):
        # Constructor parameter
        class_parameter: list[str] = []
        constructor = class_data["constructor"]
        if constructor:
            for parameter_id in constructor["parameters"]:
                parameter_data = get_data_by_id(self.api_data["parameters"], parameter_id)
                if parameter_data["name"] == "self":
                    continue
                class_parameter.append(f"{parameter_data['name']}: {parameter_data['type']['name']}")

        # Class signature line
        parameter = ", ".join(class_parameter)
        class_sign = f"class {class_data['name']}({parameter}) {{"
        f.write(class_sign)

        # Attributes
        class_attributes: list[str] = []
        for attribute_id in class_data["attributes"]:
            attribute_data = get_data_by_id(self.api_data["attributes"], attribute_id)
            if not attribute_data["is_public"]:
                continue

            attr_type = create_type_string(attribute_data["type"])
            type_string = f": {attr_type}" if attr_type else ""
            class_attributes.append(
                f"attr {attribute_data['name']}"
                f"{type_string}",
            )

        attributes = "\n\t".join(class_attributes)
        f.write(f"\n\t{attributes}")

        # Methods
        class_methods: list[str] = []
        for method_id in class_data["methods"]:
            method_data = get_data_by_id(self.api_data["functions"], method_id)
            if not method_data["is_public"]:
                continue
            class_methods.append(self._create_function_string(method_data))
        methods = "\n\n\t".join(class_methods)
        f.write(f"\n\n\t{methods}")

        # Close class
        f.write("\n}")

    def _create_function_string(self, method_data: dict) -> str:
        static = "static " if method_data["is_static"] else ""

        # Parameters
        parameters: list[str] = []
        for parameter_id in method_data["parameters"]:
            parameter_data = get_data_by_id(self.api_data["parameters"], parameter_id)
            if parameter_data["name"] == "self":
                continue
            param_type = create_type_string(parameter_data["type"])
            type_string = f": {param_type}" if param_type else ""
            parameters.append(
                f"{parameter_data['name']}"
                f"{type_string}",
            )
        method_params = ", ".join(parameters)

        # Results
        results: list[str] = []
        for result_id in method_data["results"]:
            result_data = get_data_by_id(self.api_data["results"], result_id)
            ret_type = create_type_string(result_data["type"])
            type_string = f": {ret_type}" if ret_type else ""
            results.append(
                f"{result_data['name']}"
                f"{type_string}",
            )
        method_results = f" -> {', '.join(results)}" if results else ""

        return f"{static}fun {method_data['name']}({method_params}){method_results}"

    @staticmethod
    def _write_qualified_imports(f: TextIO, qualified_imports: list) -> None:
        if not qualified_imports:
            return

        imports: list[str] = []
        for qualified_import in qualified_imports:
            qualified_name = qualified_import["qualified_name"]
            split_qname = qualified_name.split(".")
            name = split_qname.pop(-1)
            import_path = ".".join(split_qname)
            from_path = ""
            if import_path:
                from_path = f"from {import_path} "

            alias = f" as {qualified_import['alias']}" if qualified_import["alias"] else ""

            imports.append(
                f"{from_path}import {name}{alias}",
            )

        f.write(f"{'\n'.join(imports)}")

    @staticmethod
    def _write_wildcard_imports(f: TextIO, wildcard_imports: list) -> None:
        if not wildcard_imports:
            return

        imports = [
            f"from {wildcard_import['module_name']} import *"
            for wildcard_import in wildcard_imports
        ]

        f.write(f"{'\n'.join(imports)}")

    def _write_enum(self, f: TextIO, enum_data: dict) -> None:
        # Signature
        f.write(f"enum {enum_data['name']} {{\n")

        # Enum instances
        for enum_instance_id in enum_data["instances"]:
            instance = get_data_by_id(self.api_data["enum_instances"], enum_instance_id)
            f.write(f"\t{instance['name']},\n")

        # Close
        f.write("}\n\n")


def create_type_string(type_data: dict | None):
    if type_data is None:
        return ""

    kind = type_data["kind"]
    if kind == "NamedType":
        name = type_data["name"]
        match name:
            case "tuple":
                return "Tuple"
            case "int":
                return "Int"
            case "str":
                return "String"
            case "bool":
                return "Boolean"
            case "float":
                return "Float"
            case _:
                return name
    elif kind in {"OptionalType", "FinalType"}:
        # Cut out the "Type" in the kind name
        name = kind[0:-4]
        return f"{name}[{create_type_string(type_data)}]"
    elif kind in {"TupleType", "SetType", "ListType", "UnionType"}:
        types = [
            create_type_string(type_)
            for type_ in type_data["types"]
        ]
        # Cut out the "Type" in the kind name
        name = kind[0:-4]
        types_string = ""
        if types:
            types_string = f"[{', '.join(types)}]"
        return f"{name}{types_string}"
    elif kind == "DictType":
        key_data = create_type_string(type_data["key_type"])
        value_data = create_type_string(type_data["value_type"])
        key_value_data = ""
        if key_data:
            if value_data:
                key_value_data = f"[{key_data}, {value_data}]"
            else:
                key_value_data = f"[{key_data}]"
        return f"Dict{key_value_data}"
    elif kind == "LiteralType":
        return f"Literal[{', '.join(type_data['literals'])}]"

    raise ValueError(f"Unexpected type: {kind}")


def create_directory(path: Path) -> None:
    if not Path.exists(path):
        Path.mkdir(path)


def create_file(path: Path) -> None:
    Path(path).touch()


def get_data_by_id(data: list[dict], item_id: str) -> dict:
    for item in data:
        if item["id"] == item_id:
            return item
    raise ValueError("Data not found.")
