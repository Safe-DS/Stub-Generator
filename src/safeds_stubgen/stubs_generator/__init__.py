"""API-Analyzer for the Safe-DS stubs generator."""

from __future__ import annotations

from ._generate_stubs import create_stub_files, generate_stub_data
from ._helper import NamingConvention
from ._stub_string_generator import StubsStringGenerator

__all__ = [
    "create_stub_files",
    "generate_stub_data",
    "NamingConvention",
    "StubsStringGenerator",
]
