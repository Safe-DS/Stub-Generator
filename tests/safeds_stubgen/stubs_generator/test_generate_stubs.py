from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

import pytest
from safeds_stubgen.api_analyzer import TypeSourcePreference, get_api
from safeds_stubgen.docstring_parsing import DocstringStyle
from safeds_stubgen.stubs_generator import NamingConvention, StubsStringGenerator, create_stub_files, generate_stub_data

# noinspection PyProtectedMember
from safeds_stubgen.stubs_generator._generate_stubs import _convert_name_to_convention

if TYPE_CHECKING:
    from collections.abc import Generator

    from syrupy import SnapshotAssertion


# Setup - Run API to create stub files
_lib_dir = Path(__file__).parent.parent.parent
_test_package_name = "various_modules_package"
_test_package_dir = Path(_lib_dir / "data" / _test_package_name)
_out_dir = Path(_lib_dir / "data" / "out")
_out_dir_stubs = Path(_out_dir / "tests/data" / _test_package_name)

_docstring_package_name = "docstring_parser_package"
_docstring_package_dir = Path(_lib_dir / "data" / _docstring_package_name)

api = get_api(_test_package_dir, is_test_run=True)
stubs_generator = StubsStringGenerator(api=api, convert_identifiers=True)
stubs_data = generate_stub_data(stubs_generator=stubs_generator, out_path=_out_dir)


def test_file_creation() -> None:
    data_to_test: list[tuple[str, str]] = [
        ("/".join(stub_data[0].parts), stub_data[1]) for stub_data in stubs_data if "file_creation" in str(stub_data[0])
    ]
    data_to_test.sort(key=lambda x: x[1])

    expected_files: list[tuple[str, str]] = [
        ("tests/data/various_modules_package/file_creation/Lv1", "Lv1"),
        (
            "tests/data/various_modules_package/file_creation/ReexportedInAnotherPackageClass",
            "ReexportedInAnotherPackageClass",
        ),
        (
            "tests/data/various_modules_package/file_creation/ReexportedInAnotherPackageClass2",
            "ReexportedInAnotherPackageClass2",
        ),
        ("tests/data/various_modules_package/file_creation/module_1", "module_1"),
        ("tests/data/various_modules_package/file_creation/package_1/module_5", "module_5"),
        ("tests/data/various_modules_package/file_creation/public_reexported", "public_reexported"),
        ("tests/data/various_modules_package/file_creation", "reexported_from_another_package_3"),
        (
            "tests/data/various_modules_package/file_creation/reexported_in_another_package_function",
            "reexported_in_another_package_function",
        ),
        (
            "tests/data/various_modules_package/file_creation/reexported_in_another_package_function2",
            "reexported_in_another_package_function2",
        ),
    ]
    expected_files.sort(key=lambda x: x[1])

    assert len(data_to_test) == len(expected_files)
    for data_tuple, expected_tuple in zip(data_to_test, expected_files, strict=True):
        assert data_tuple[0].endswith(expected_tuple[0])
        assert data_tuple[1] == expected_tuple[1]


def test_file_creation_limited_stubs_outside_package(snapshot_sds_stub: SnapshotAssertion) -> None:
    create_stub_files(stubs_generator=stubs_generator, stubs_data=stubs_data, out_path=_out_dir)

    path = Path(_out_dir / "tests/data/main_package/another_path/another_module/another_module.sdsstub")
    assert path.is_file()

    with path.open("r", encoding="utf-8") as f:
        assert f.read() == snapshot_sds_stub


def stub_texts() -> Generator[str, None, None]:
    for stub_data in stubs_data:
        yield stub_data[2]


def _stub_ids() -> Generator[str, None, None]:
    for stub_data in stubs_data:
        yield stub_data[1]


@pytest.mark.parametrize("stub_text", stub_texts(), ids=_stub_ids())
class TestStubFileGeneration:
    def test_stub_creation(self, stub_text: str, snapshot_sds_stub: SnapshotAssertion) -> None:
        assert stub_text == snapshot_sds_stub


@pytest.mark.parametrize(
    ("name", "expected_result", "naming_convention", "is_class_name"),
    [
        ("", "", "Safe-DS", False),
        ("_", "_", "Safe-DS", False),
        ("__get_function_name__", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("__get_function_name", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("get_function_name__", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("__getFunction_name__", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("__get__function___name__", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("__get_funCtion_NamE__", "getFunCtionNamE", NamingConvention.SAFE_DS, False),
        ("getFunctionName", "getFunctionName", NamingConvention.SAFE_DS, False),
        ("a_a_A_aAAaA_1_1_2_aAa", "aAAAAAaA112AAa", NamingConvention.SAFE_DS, False),
        ("some_class_name", "SomeClassName", NamingConvention.SAFE_DS, True),
        ("some_class_name", "some_class_name", NamingConvention.PYTHON, True),
        ("__get_function_name__", "__get_function_name__", NamingConvention.PYTHON, False),
    ],
)
def test_convert_name_to_convention(
    name: str,
    expected_result: str,
    naming_convention: NamingConvention,
    is_class_name: bool,
) -> None:
    assert (
        _convert_name_to_convention(name=name, naming_convention=naming_convention, is_class_name=is_class_name)
        == expected_result
    )


@pytest.mark.parametrize(
    ("filename", "docstring_style", "type_source_preference"),
    [
        ("full_docstring", DocstringStyle.PLAINTEXT, TypeSourcePreference.CODE),
        ("googledoc", DocstringStyle.GOOGLE, TypeSourcePreference.CODE),
        ("numpydoc", DocstringStyle.NUMPYDOC, TypeSourcePreference.CODE),
        ("plaintext", DocstringStyle.PLAINTEXT, TypeSourcePreference.CODE),
        ("restdoc", DocstringStyle.REST, TypeSourcePreference.CODE),
        ("docstring_vs_typehints", DocstringStyle.NUMPYDOC, TypeSourcePreference.CODE),
        ("docstring_vs_typehints", DocstringStyle.NUMPYDOC, TypeSourcePreference.DOCSTRING),
    ],
    ids=[
        "full_docstring-PLAINTEXT",
        "googledoc-GOOGLE",
        "numpydoc-NUMPYDOC",
        "plaintext-PLAINTEXT",
        "restdoc-REST",
        "docstring_vs_typehints-CODE",
        "docstring_vs_typehints-DOCSTRING",
    ],
)
def test_stub_docstring_creation(
    filename: str,
    docstring_style: DocstringStyle,
    type_source_preference: TypeSourcePreference,
    snapshot_sds_stub: SnapshotAssertion,
) -> None:
    docstring_api = get_api(
        root=_docstring_package_dir,
        docstring_style=docstring_style,
        is_test_run=True,
        type_source_preference=type_source_preference,
    )
    docstring_stubs_generator = StubsStringGenerator(api=docstring_api, convert_identifiers=True)
    docstring_stubs_data = generate_stub_data(stubs_generator=docstring_stubs_generator, out_path=_out_dir)

    for stub_text in docstring_stubs_data:
        if stub_text[1] == filename:
            assert stub_text[2] == snapshot_sds_stub
            return

    raise pytest.fail(f"Could not find data for '{filename}'.")
