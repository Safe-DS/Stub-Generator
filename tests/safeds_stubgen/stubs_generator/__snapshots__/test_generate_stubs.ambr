# serializer version: 1
# name: test_abstract_creation
  '''
  @PythonModule("various_modules_package.abstract_module")
  package variousModulesPackage.abstractModule
  
  from abc import ABC
  from abc import abstractmethod
  
  class AbstractModuleClass {
  	@PythonName("abstract_property_method") attr abstractPropertyMethod: union<Float, Int>
  
  	// TODO Result type information missing.
  	@Pure
  	@PythonName("abstract_method")
  	fun abstractMethod()
  
  	// TODO List type has to many type arguments.
  	@Pure
  	@PythonName("abstract_method_params")
  	fun abstractMethodParams(
  		@PythonName("param_1") param1: Int,
  		@PythonName("param_2") param2: Boolean = false
  	) -> result1: List<String, Int>
  
  	// TODO Result type information missing.
  	@Pure
  	@PythonName("abstract_static_method")
  	static fun abstractStaticMethod()
  
  	@Pure
  	@PythonName("abstract_static_method_params")
  	static fun abstractStaticMethodParams(
  		param: Float
  	) -> result1: Boolean
  }
  
  '''
# ---
# name: test_class_attribute_creation
  '''
  @PythonModule("various_modules_package.attribute_module")
  package variousModulesPackage.attributeModule
  
  from typing import Optional
  from typing import Final
  from typing import Literal
  
  class AttributesClassA()
  
  class AttributesClassB() {
  	@PythonName("type_hint_public")
  	static attr typeHintPublic: Int
  	@PythonName("no_type_hint_public")
  	static attr noTypeHintPublic: Int
  	@PythonName("object_attr")
  	static attr objectAttr: AttributesClassA
  	// TODO Attribute has no type information.
  	@PythonName("callexpr_attr_class")
  	static attr callexprAttrClass
  	// TODO Attribute has no type information.
  	@PythonName("callexpr_attr_function")
  	static attr callexprAttrFunction
  	// TODO Safe-DS does not support tuple types.
  	@PythonName("tuple_attr_1")
  	static attr tupleAttr1: Tuple<Any>
  	// TODO Safe-DS does not support tuple types.
  	@PythonName("tuple_attr_2")
  	static attr tupleAttr2: Tuple<union<Int, String>>
  	// TODO Safe-DS does not support tuple types.
  	@PythonName("tuple_attr_3")
  	static attr tupleAttr3: Tuple<String, Int>
  	@PythonName("list_attr_1")
  	static attr listAttr1: List<Any>
  	@PythonName("list_attr_2")
  	static attr listAttr2: List<union<AttributesClassA, String>>
  	// TODO List type has to many type arguments.
  	@PythonName("list_attr_3")
  	static attr listAttr3: List<String, AttributesClassA>
  	// TODO List type has to many type arguments.
  	@PythonName("list_attr_4")
  	static attr listAttr4: List<String, union<AttributesClassA, Int>>
  	@PythonName("set_attr_1")
  	static attr setAttr1: Set<Any>
  	@PythonName("set_attr_2")
  	static attr setAttr2: Set<union<AttributesClassA, String>>
  	// TODO Set type has to many type arguments.
  	@PythonName("set_attr_3")
  	static attr setAttr3: Set<String, AttributesClassA>
  	// TODO Set type has to many type arguments.
  	@PythonName("set_attr_4")
  	static attr setAttr4: Set<String, union<AttributesClassA, Int>>
  	@PythonName("dict_attr_1")
  	static attr dictAttr1: Map<Any, Any>
  	@PythonName("dict_attr_2")
  	static attr dictAttr2: Map<String, Int>
  	@PythonName("dict_attr_3")
  	static attr dictAttr3: Map<union<Int, String>, AttributesClassA?>
  	@PythonName("bool_attr")
  	static attr boolAttr: Boolean
  	@PythonName("none_attr")
  	static attr noneAttr: Nothing?
  	@PythonName("flaot_attr")
  	static attr flaotAttr: Float
  	@PythonName("int_or_bool_attr")
  	static attr intOrBoolAttr: union<Boolean, Int>
  	@PythonName("str_attr_with_none_value")
  	static attr strAttrWithNoneValue: String
  	static attr optional: Int?
  	static attr final: String
  	static attr finals: union<Int, String>
  	@PythonName("final_union")
  	static attr finalUnion: union<Int, String>
  	static attr `literal`: literal<"Some String">
  	@PythonName("multi_attr_1")
  	static attr multiAttr1: Int
  	@PythonName("multi_attr_3")
  	static attr multiAttr3: List<String>
  	@PythonName("multi_attr_5")
  	static attr multiAttr5: String
  	@PythonName("multi_attr_6")
  	static attr multiAttr6: String
  	@PythonName("multi_attr_7")
  	static attr multiAttr7: String
  	@PythonName("multi_attr_8")
  	static attr multiAttr8: String
  	@PythonName("init_attr")
  	attr initAttr: Boolean
  
  	@Pure
  	@PythonName("some_func")
  	static fun someFunc() -> result1: Boolean
  }
  
  '''
# ---
# name: test_class_creation
  '''
  @PythonModule("various_modules_package.class_module")
  package variousModulesPackage.classModule
  
  class ClassModuleEmptyClassA()
  
  class ClassModuleClassB(
  	a: Int,
  	b: ClassModuleEmptyClassA?
  ) sub ClassModuleEmptyClassA {
  	// TODO Result type information missing.
  	@Pure
  	fun f()
  }
  
  // TODO Safe-DS does not support multiple inheritance.
  class ClassModuleClassC() sub ClassModuleEmptyClassA, ClassModuleClassB {
  	@PythonName("attr_1")
  	static attr attr1: Int
  	@PythonName("attr_2")
  	static attr attr2: Int
  
  	// TODO Result type information missing.
  	@Pure
  	fun f1()
  }
  
  class ClassModuleClassD() {
  	class ClassModuleNestedClassE() {
  		@PythonName("nested_attr_1")
  		static attr nestedAttr1: Nothing?
  
  		@PythonName("_ClassModulePrivateDoubleNestedClassF")
  		class ClassModulePrivateDoubleNestedClassF()
  
  		// TODO Result type information missing.
  		@Pure
  		@PythonName("class_e_func")
  		fun classEFunc()
  	}
  }
  
  '''
# ---
# name: test_enum_creation
  '''
  @PythonModule("various_modules_package.enum_module")
  package variousModulesPackage.enumModule
  
  from another_path.another_module import AnotherClass as _AcImportAlias
  
  enum _ReexportedEmptyEnum
  
  enum EnumTest {
  	ONE
  	TWO
  	THREE
  	FOUR
  	FIVE
  	SIX
  	SEVEN
  	EIGHT
  	NINE
  }
  
  enum EnumTest2 {
  	TEN
  }
  
  enum EnumTest3 {
  	ELEVEN
  }
  
  enum EmptyEnum
  
  '''
# ---
# name: test_function_creation
  '''
  @PythonModule("various_modules_package.function_module")
  package variousModulesPackage.functionModule
  
  from typing import Callable
  from typing import Optional
  from typing import Literal
  from typing import Any
  
  // TODO Result type information missing.
  @Pure
  @PythonName("public_no_params_no_result")
  fun publicNoParamsNoResult()
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support tuple types.
  // TODO Some parameter have no type information.
  @Pure
  fun params(
  	integer: Int,
  	boolean: Boolean,
  	@PythonName("float_") float: Float,
  	none: Nothing?,
  	string: String,
  	obj: FunctionModuleClassA,
  	callexpr,
  	`union`: union<Boolean, Int>,
  	@PythonName("list_") list: List<Int>,
  	dictionary: Map<String, union<Float, Int>>,
  	@PythonName("set_") set: Set<String>,
  	optional: Int?,
  	@PythonName("tuple_") tuple: Tuple<Int, String, Boolean>,
  	`literal`: literal<"Some String">,
  	@PythonName("any_") any: Any
  )
  
  // TODO List type has to many type arguments.
  // TODO Result type information missing.
  // TODO Safe-DS does not support tuple types.
  @Pure
  @PythonName("illegal_params")
  fun illegalParams(
  	lst: List<Int, String>,
  	@PythonName("lst_2") lst2: List<Int, String, Int>,
  	tpl: Tuple<Int, String, Boolean, Int>,
  	`_`: Int = String
  )
  
  // TODO Result type information missing.
  @Pure
  @PythonName("special_params")
  fun specialParams(
  	@PythonName("none_union") noneUnion: Nothing?,
  	@PythonName("none_bool_union") noneBoolUnion: Boolean?,
  	@PythonName("bool_none_union") boolNoneUnion: Boolean?,
  	@PythonName("none_bool_none_union") noneBoolNoneUnion: Boolean?,
  	@PythonName("none_bool_int_union") noneBoolIntUnion: union<Boolean, Int, Nothing?>,
  	@PythonName("none_none_bool_none_union") noneNoneBoolNoneUnion: Boolean?,
  	@PythonName("none_list_union_none_none") noneListUnionNoneNone: List<Nothing?>?,
  	none: Nothing?
  )
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support required but name only parameter assignments.
  // TODO Some parameter have no type information.
  @Pure
  @PythonName("param_position")
  fun paramPosition(
  	self,
  	a,
  	b: Boolean,
  	c: FunctionModuleClassA = FunctionModuleClassA,
  	d,
  	e: Int = 1
  )
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support optional but position only parameter assignments.
  // TODO Some parameter have no type information.
  @Pure
  @PythonName("opt_pos_only")
  fun optPosOnly(
  	required,
  	optional: Int = 1
  )
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support required but name only parameter assignments.
  // TODO Some parameter have no type information.
  @Pure
  @PythonName("req_name_only")
  fun reqNameOnly(
  	required,
  	optional: Int = 1
  )
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support variadic parameters.
  // TODO Some parameter have no type information.
  @Pure
  fun arg(
  	args: List<Any>,
  	kwargs: Map<String, Any>
  )
  
  // TODO Result type information missing.
  // TODO Safe-DS does not support variadic parameters.
  @Pure
  @PythonName("args_type")
  fun argsType(
  	args: List<Int>,
  	kwargs: Map<String, Int>
  )
  
  @Pure
  @PythonName("int_result")
  fun intResult() -> result1: Int
  
  @Pure
  @PythonName("str_result")
  fun strResult() -> result1: String
  
  @Pure
  @PythonName("bool_result")
  fun boolResult() -> result1: Boolean
  
  @Pure
  @PythonName("float_result")
  fun floatResult() -> result1: Float
  
  // TODO Result type information missing.
  @Pure
  @PythonName("none_result")
  fun noneResult()
  
  @Pure
  @PythonName("obj_result")
  fun objResult() -> result1: FunctionModuleClassA
  
  // TODO Result type information missing.
  @Pure
  @PythonName("callexr_result_class")
  fun callexrResultClass()
  
  // TODO Result type information missing.
  @Pure
  @PythonName("callexr_result_function")
  fun callexrResultFunction()
  
  @Pure
  @PythonName("tuple_results")
  fun tupleResults() -> (result1: String, result2: FunctionModuleClassA)
  
  @Pure
  @PythonName("union_results")
  fun unionResults() -> result1: union<Int, String>
  
  @Pure
  @PythonName("list_results")
  fun listResults() -> result1: List<Int>
  
  // TODO List type has to many type arguments.
  @Pure
  @PythonName("illegal_list_results")
  fun illegalListResults() -> result1: List<Int, String>
  
  @Pure
  @PythonName("dictionary_results")
  fun dictionaryResults() -> result1: Map<String, FunctionModuleClassA>
  
  @Pure
  @PythonName("illegal_dictionary_results")
  fun illegalDictionaryResults() -> result1: Map<Any, Any>
  
  @Pure
  @PythonName("union_dictionary_results")
  fun unionDictionaryResults() -> result1: Map<union<Int, String>, union<Boolean, Float>>
  
  @Pure
  @PythonName("set_results")
  fun setResults() -> result1: Set<String>
  
  // TODO Set type has to many type arguments.
  @Pure
  @PythonName("illegal_set_results")
  fun illegalSetResults() -> result1: Set<String, Boolean>
  
  @Pure
  @PythonName("optional_results")
  fun optionalResults() -> result1: Int?
  
  @Pure
  @PythonName("literal_results")
  fun literalResults() -> result1: literal<"Some String">
  
  @Pure
  @PythonName("any_results")
  fun anyResults() -> result1: Any
  
  @Pure
  @PythonName("callable_type")
  fun callableType(
  	param: (a: String) -> (b: Int, c: String)
  ) -> result1: (a: Int, b: Int) -> c: Int
  
  class FunctionModuleClassA()
  
  // TODO Some parameter have no type information.
  class FunctionModuleClassB(
  	@PythonName("init_param") initParam
  ) {
  	class FunctionModuleClassC() {
  		class FunctionModuleClassD()
  
  		@Pure
  		@PythonName("nested_class_function")
  		fun nestedClassFunction(
  			param1: Int
  		) -> result1: Boolean
  	}
  
  	@Pure
  	@PythonName("instance_method")
  	fun instanceMethod(
  		a: FunctionModuleClassA
  	) -> result1: FunctionModuleClassA
  
  	// TODO Result type information missing.
  	@Pure
  	@PythonName("static_method")
  	static fun staticMethod()
  
  	// TODO Result type information missing.
  	@Pure
  	@PythonName("static_method_params")
  	static fun staticMethodParams(
  		@PythonName("param_1") param1: Int
  	)
  
  	// TODO Result type information missing.
  	// TODO Safe-DS does not support class methods.
  	@Pure
  	@PythonName("class_method")
  	static fun classMethod()
  
  	// TODO Safe-DS does not support class methods.
  	@Pure
  	@PythonName("class_method_params")
  	static fun classMethodParams(
  		@PythonName("param_1") param1: Int
  	) -> result1: Boolean
  }
  
  class FunctionModulePropertiesClass() {
  	@PythonName("property_function") attr propertyFunction
  	@PythonName("property_function_params") attr propertyFunctionParams: String
  	@PythonName("property_function_infer") attr propertyFunctionInfer: String
  }
  
  '''
# ---
# name: test_import_creation
  '''
  @PythonModule("various_modules_package.import_module")
  package variousModulesPackage.importModule
  
  import mypy as `static`
  
  from math import *
  
  '''
# ---
# name: test_type_inference
  '''
  @PythonModule("various_modules_package.infer_types_module")
  package variousModulesPackage.inferTypesModule
  
  class InferMe()
  
  class InferMe2()
  
  class InferMe3()
  
  class InferMyTypes(
  	@PythonName("init_param") initParam: Int = 1
  ) {
  	@PythonName("infer_int")
  	static attr inferInt: Int
  	@PythonName("infer_float")
  	static attr inferFloat: Float
  	@PythonName("infer_bool")
  	static attr inferBool: Boolean
  	@PythonName("infer_str")
  	static attr inferStr: String
  	@PythonName("infer_none")
  	static attr inferNone: Nothing?
  	@PythonName("infer_obj")
  	static attr inferObj: () -> a: InferMe
  	// TODO Attribute has no type information.
  	@PythonName("init_infer")
  	attr initInfer
  
  	@Pure
  	@PythonName("infer_function")
  	static fun inferFunction(
  		@PythonName("infer_param") inferParam: Int = 1,
  		@PythonName("infer_param_2") inferParam2: Int = Something
  	) -> (result1: union<Boolean, Float, InferMe, InferMe2, InferMe3, InferMyTypes, Int, Nothing?, String>, result2: union<Float, Int>, result3: Float)
  }
  
  '''
# ---
# name: test_variance_creation
  '''
  @PythonModule("various_modules_package.variance_module")
  package variousModulesPackage.varianceModule
  
  from typing import Generic
  from typing import TypeVar
  from typing import Literal
  
  class A()
  
  class VarianceClassAll<out TCo, in TCon, TIn>() where {
  	TCo sub String,
  	TCon super A
  }
  
  class VarianceClassOnlyInvariance<TIn>()
  
  '''
# ---
